opm/simulators/linalg/cuistl/PreconditionerConvertFieldTypeAdapter.hpp

```diff
@@ -0,0 +1,240 @@
+//!
+//! template<class ParallelInfo>
+//! void applyILU0AsFloat(const MDouble& matrix, const XDouble& x, XDouble& y) {
+//!
+//!     using FloatILU0 = typename Opm::ParallelOverlappingILU0<MFloat, XFloat, XFloat, ParallelInfo>;
+//!     using DoubleToFloatConverter = typename Opm::cuistl::PreconditionerConvertFieldTypeAdapter<FloatILU0, MDouble,
+//!     XDouble, XDouble>;
+//!
+//!     // Note that we do not need to make a new instance for every invocation, this
+//!     // is just done for this example
+//!     auto doubleToFloatConverter = DoubleToFloatConverter(matrix);
+//!     const auto& convertedMatrix = doubleToFloatConverter.getConvertedMatrix();
+//!
+//!     auto floatILU0 = std::make_shared<FloatILU0>(convertedMatrix, 0, 1.0, 0);
+//!
+//!     doubleToFloatConverter.setUnderlyingPreconditioner(floatILU0);
+//!
+//!     // This will convert x and y to float, thenn call floatILU0.apply on the converted arguments
+//!     doubleToFloatConverter.apply(x, y);
+//! }
+//!
+//! \endcode
+template <class CudaPreconditionerType, class M, class X, class Y, int l = 1>
+class PreconditionerConvertFieldTypeAdapter : public Dune::PreconditionerWithUpdate<X, Y>
+{
+public:
+    //! \brief The matrix type the preconditioner is for.
+    using matrix_type = typename std::remove_const<M>::type;
+
+    //! \brief The domain type of the preconditioner.
+    using domain_type = X;
+    //! \brief The range type of the preconditioner.
+    using range_type = Y;
+    //! \brief The field type of the preconditioner.
+    using field_type = typename X::field_type;
+
+
+    using domain_type_to = typename CudaPreconditionerType::domain_type;
+    //! \brief The range type of the preconditioner.
+    using range_type_to = typename CudaPreconditionerType::range_type;
+    //! \brief The field type of the preconditioner.
+    using field_type_to = typename domain_type_to::field_type;
+
+
+    using block_type = typename domain_type::block_type;
+
+    using XTo = Dune::BlockVector<Dune::FieldVector<field_type_to, block_type::dimension>>;
+    using YTo = Dune::BlockVector<Dune::FieldVector<field_type_to, block_type::dimension>>;
+    using matrix_type_to =
+        typename Dune::BCRSMatrix<Dune::FieldMatrix<field_type_to, block_type::dimension, block_type::dimension>>;
+
+    //! \brief Constructor.
+    //!
+    //! \param A The matrix to operate on.
+    //!
+    //! \note After the PreconditionerConvertFieldTypeAdapter you can get the converted matrix
+    //! by calling getConvertedMatrix(), which in turn can be used to create the underlying preconditioner.
+    //! Once the underlying precondtioner has been called, this must be supplied to setUnderlyingPreconditioner.
+    explicit PreconditionerConvertFieldTypeAdapter(const M& matrix)
+        : m_matrix(matrix)
+        , m_convertedMatrix(createConvertedMatrix())
+    {
+    }
+
+    //! \brief Not used at the moment
+    virtual void pre([[maybe_unused]] X& x, [[maybe_unused]] Y& b) override
```

Comment:
Should this not simply call the underlying preconditioner's pre() function (after the required conversions)? To avoid calling when not needed it should implement shouldCallPre() to forward the underlying preconditioner's behaviour. As is written here, I think this prevents me from testing arbitrary CPU preconditioners with float, which would be useful for experiments.

Similar for post(). Also the same applies to the PreconditionerAdapter class.

Reply:
It was to avoid adding more conversion code when the function is not at all used by ILU0, for which it was tested. I suggest adding this when the need arise.